subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/FileSystemSnapshotter.java

```diff
@@ -49,7 +48,16 @@
     FileSystemLocationSnapshot snapshot(File file);
 
     /**
-     * Returns snapshots of the roots of a file collection.
+     * Snapshots a directory tree.
+     *
+     * For simplicity this only caches trees without includes/excludes. However, if it is asked
+     * to snapshot a filtered tree, it will try to find a snapshot for the underlying
+     * tree and filter it in memory instead of walking the file system again. This covers the
+     * majority of cases, because all task outputs are put into the cache without filters
+     * before any downstream task uses them.
+     *
+     * If it turns out that a filtered tree has actually not been filtered (i.e. the condition always returned true),
+     * then we cache the result as unfiltered tree.
      */
-    List<FileSystemSnapshot> snapshot(FileCollectionInternal fileCollection);
+    FileSystemSnapshot snapshotDirectoryTree(File root, PatternSet patterns);
```

Comment:
Why did we return a `List` before? :o

Reply:
The diff is misleading: The old method snapshotted a file collection. Therefore, it returned a list of snapshots. The new method snapshots a single directory tree. So it only returns one snapshot.